### ISSUE TITLE
When adding SSL to my web server, do I also need to carry out additional steps in my local VVV installation?

### ISSUE THREAD

**Issue by cagross** [Varying-Vagrant-Vagrants/VVV#1433]
## Expected Behavior
N/A

## Current Behavior
N/A

## Possible Solution

A web search has returned several posts on the matter.  But none appear to be from official VVV documentation.  I'd like to know the current recommendation in terms of SSL.

## Context
I have VVV installed on my local machine, and I use it to develop my WordPress site, then push edits to my remove web server.  To the remote website, I plan on adding an SSL certificate (using [these instructions](https://uk.godaddy.com/help/install-a-lets-encrypt-ssl-apache-20245)).  As part of this process, do I need to carry out any additional steps on the local VVV side?  If so, what exactly do I need to do?

## Your Environment

* VVV version:  2.1.0-73-g6a86225 
* VVV Git Branch:   master
* Vagrant version: 2.0.1
* VM Provider name:   VirtualBox
* VM Provider version:  5.2.2 r119230 (Qt5.6.2)
* Operating System and version:  Win `10, x64.

**Comment by tomjn** [comment]
Right now there's no recommended instructions, hence why there's no SSL
page in the docs. It's possible to do, but the specifics have yet to be
listed in a single place and confirmed by a VVV user

I can tell you though with absolute certainty that the godaddy article you
linked to will not work, and for several reasons:

- that's for Apache, VVV uses nginx
- let's encrypt doesn't do local dev environments. Part of the protocol
involves let's encrypt contacting your site to confirm it is who it says it
is, you'd have to run VVV exposed on the open web, running on the domain
you want to test. Impossible if you're using a test TLD, and it'd need VVV
running in production

You need a self signed certificate. Ideally from a self created certificate
authority. You also need to modify the nginx configs to use that
certificate. It won't work with .Dev domains ( because Google ), and it
won't work with the default sites ( because of a certificate generated for
legacy reasons that needs removing/changing )

**Comment by cagross** [comment]
OK thanks for the reply.  I hear what you are saying.  But are you able to point me to any links that might help me add SSL to my VVV site--maybe one that has helped you in the past?  I would imagine many people before me would have encountered this issue.

As an alternative, would I be able to add the SSL certificate to my remote site, yet leave it unchanged on my VVV site?  Would there be any issues if I did that?  One potential issue I see with that is in the database sync between the two sites. Namely, if I migrate to SSL on my web server, part of the process would involve a global search and replace of the database on the remote site (i.e. roughly speaking, I would replace all `http` strings with `https`).  But I would just lose these changes after the next VVV edit, since the edit would entail pushing my local VVV database to my remote database.  I'm not sure how to resolve that.  Would I be able to carry out the same `http`-->`https` search and replace on my VVV site?  If so, then I guess the two databases would remain in-sync, alleviating the issue I described.

> I can tell you though with absolute certainty that the godaddy article you
linked to will not work,

Gotcha.  I wasn't suggesting that those instructions would work for my VVV installation--I was reasonably sure they wouldn't. I was just disclosing what method I would use on my web server.  But good to know the reasons why they don't apply to VVV.

**Comment by tomjn** [comment]
> OK thanks for the reply. I hear what you are saying. But are you able to point me to any links that might help me add SSL to my VVV site--maybe one that has helped you in the past? I would imagine many people before me would have encountered this issue.

I don't have a reference I can hand to you, I've personally not tried to follow any in a long time so I can't vouch for an article. But if I did you'd be sure I'd paraphrase it into an SSL page under references on the docs

> As an alternative, would I be able to add the SSL certificate to my remote site, yet leave it unchanged on my VVV site? Would there be any issues if I did that?

You definitely should switch to https for your production site, and I would expect you could download the certificate letsencrypt gives you then modify `vvv-nginx.conf`/`vvv-nginx.conf.tmpl` to use that certificate, I would expect that to work. Of course it would need to be an additional site using the custom site template, rather than the ones that are setup when you first provision VVV ( those are under the default SSL certificate, which needs removing ). There's a security risk of doing that, but if you're sure your local environment is secure from prying eyes and isn't shared beyond your machine then that should be ok ( and you'll need to re-download the certificate regularly, letsencrypt certificates don't last long )

> One potential issue I see with that is in the database sync between the two sites. Namely, if I migrate to SSL on my web server, part of the process would involve a global search and replace of the database on the remote site (i.e. roughly speaking, I would replace all http strings with https). But I would just lose these changes after the next VVV edit, since the edit would entail pushing my local VVV database to my remote database. I'm not sure how to resolve that.

Is a WXR import/export an option? I see the issue here, SSL is indeed on the agenda, a lot of people have popped up the last 2-3 weeks asking about it, so hopefully this can be wrangled soonish

> Would I be able to carry out the same http-->https search and replace on my VVV site? If so, then I guess the two databases would remain in-sync, alleviating the issue I described.

There's all the usual methods available inside the VVV VM, I'd suggest WP CLI search replace

**Comment by cagross** [comment]
> Is a WXR import/export an option?

I'm not sure what WXR import/export is.  I've never heard of it :-/

>> There's all the usual methods available inside the VVV VM, I'd suggest WP CLI search replace

Well what I was getting at when asking about the search and replace was whether I could get away with updating SSL on my remote site, yet foregoing it on my local VVV site.  My feeling though is that will not be possible, since the remote site (files, database) will have their URLs updated from `http` to `https`, while the local VVV site will not.  Would you agree that this update-remote-while-forego-on-local option would not be viable?  In other words, if I migrate to SSL on my remote site, I will have to do the same on my local site--correct?  Or have I misunderstood something?

**Comment by tomjn** [comment]
I don't know how viable converting from https to http for local then back again via search replace to https is

> I'm not sure what WXR import/export is. I've never heard of it :-/

The WP importer/exporter generates WXR files

**Comment by cagross** [comment]
> I don't know how viable converting from https to http for local then back again via search replace to https is. It's probably doable, but it's not something I've had to do before

OK, it sounds like I'm not explaining this method well enough.   My mistake, as I probably don't understand the underlying concepts of SSL as much as I should.  But just to be clear, let me try again.

Let's say I've setup a site on my local computer with VVV, and I've successfully pushed it (files and database) to my remote server.  Then I successfully install an SSL certificate on my _remote server only_, and ensure everything in WordPress is SSL compliant (i.e. URLs are changed from `http` to `https`, etc).  At the same time, I leave my local VVV site _as-is_ as far as SSL.  If I then make a new edit locally, and push the edit (files and database) to my remote server, will there be new SSL-related issues on my remote site?  My intuition says 'yes,' because in the very least it sounds like I will be pushing the `http` URLs from my local site to my SSL-enabled remote site, which requires `https` URLs.  Am I correct here, or have I misunderstood something?

I'd just like to be concrete on this method before continuing to discuss SSL methods for VVV.  I am very happy to discuss those methods further, and would prefer to do so, in this thread, if you are willing to discuss further.  Also, if you want to take this opportunity to use me as a guinea pig so to speak, I would be willing.   If you wanted to describe a rough method of installing SSL for VVV, I could do so, document my steps, and go through some trial and error with you until I get it working.  That way maybe you could at least have the start of some SSL-for-VVV documentation.  Of course I'd only be able to test on my single system (Windows 10).  Anyway, just throwing that out there.

**Comment by tomjn** [comment]
@cagross changes have been made:

 - Update to the latest VVV `develop`
 - Add `tls-ca` under `core:` to the utilities section, [as shown here](https://github.com/Varying-Vagrant-Vagrants/VVV/blob/develop/vvv-config.yml#L59)
 - Reprovision

If you're using the wordpress-develop or custom site templates, visit your site and switch to `https`. The first time you visit it will complain, but you should be able to skip.

Once reprovisioned, you can also go to `certificates/ca/cert.crt`. Add this to your keychain on MacOS/Accept the certificate as an authority in Windows/etc, and restart your browser. Now all VVV sites using the custom site template support https, with no warnings

**Comment by cagross** [comment]
OK thanks very much for that.  I'll try to test those instructions tonight (hopefully within ~12-15 hours).  I'll leave feedback here if that's OK.

**Comment by tomjn** [comment]
WIP docs here https://github.com/Varying-Vagrant-Vagrants/varyingvagrantvagrants.org/pull/73

**Comment by cagross** [comment]
I just want to confirm a few things before I start:

I have not begun to install an SSL certificate on my production site, and my VVV site and production site are completely in-sync.  Should I do migrate my production site to SSL separately, before carrying out your instructions?  Or in-theory, should it not matter which I do first?

>Once reprovisioned, you can also go to certificates/ca/cert.crt. Add this to your keychain on MacOS/Accept the certificate as an authority in Windows/etc, and restart your browser. Now all VVV sites using the custom site template support https, with no warnings

This is a required step, not a merely suggested step--correct?  Not complaining or anything, but I just want to be sure.

Thanks.

**Comment by tomjn** [comment]
@cagross this just makes VVV serve valid certificates, so you can use `https`, your WP site will still generate `http` URLs for links etc, afterall a certificate authority does not reach into WordPress and change the home and site URLs from http to https. Same way that handing you a secure phone line allows you to talk about secret stuff, but it's no guarantee that's what you're going to talk about.

Beyond this point, there's no difference between VVV and any other local environment, or server to server migration from your point of view, but my recommendation is to use this and verify it loads via https for you, then switch all your sites over.

**Comment by tomjn** [comment]
> This is a required step, not a merely suggested step--correct? Not complaining or anything, but I just want to be sure.

If you don't do this, every new site you provision will throw up a security error when you first open them, but you can dismiss it and it won't show again in that browser

If you do it though, it will never happen and any new site will just work automatically.

**Comment by cagross** [comment]
@tomjn OK thanks for explaining that.

I am starting your instructions now.  The first step is:

> Update to the latest VVV `develop`

How exactly can I confirm if I am running VVV `develop`, and how would I in-turn update to the latest version?

**Comment by tomjn** [comment]
You can switch to the `develop` branch as you would any other git repo like this:

```
git checkout develop
git pull
```

**Comment by cagross** [comment]
Ah OK, so `develop` is my current branch--got it.

OK I'm already on the `develop` branch, and carried out a `git pull` without issue.

The next step is:

> Add tls-ca under core: to the utilities section,

The link you posted shows how to add `tls-ca` to `vvv-config.yml`.  But my `vvv-config.yml` already has this entry.  However my `vvv-custom.yml` does not.  Should I add the `tls-ca` line to `vvv-custom.yml`?  Or should I move on to the next step (reprovision)?

**Comment by tomjn** [comment]
Yeah, `vvv-custom.yml`, `vvv-config.yml` is just the default

**Comment by cagross** [comment]
OK thanks.  I carried out that step without issue.  I then carried out the 'Reprovision' step by executing a `vagrant up`, then a `vagrant provision`.  The `provision` step returned some warnings/errors regarding the database. The output in-question is:

```
  default: Creating database 'cagross' (if it's not already there)
    default:  DB operations done.
    default: PHP Notice:  Undefined index: SERVER_NAME in phar:///usr/local/bin/wp/php/WP_CLI/Runner.php(1138) : eval()'d code on line 22
    default: PHP Warning:  mysqli_real_connect(): (HY000/1045): Access denied for user 'db-user-el'@'localhost' (using password: YES) in /srv/www/cagross/public_html/wp-includes/wp-db.php on line 1531
    default: Warning: mysqli_real_connect(): (HY000/1045): Access denied for user 'db-user-el'@'localhost' (using password: YES) in /srv/www/cagross/public_html/wp-includes/wp-db.php on line 1531
    default: Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or we can’t contact the database server at `localhost`. This could mean your host’s database server is down.

```
This seems to indicate that a connection cannot be made to my database.  But I can still open the site in-question (http://cagross.test), both the front-end and back-end.  Shouldn't that be impossible if there is an error establishing a database connection?

I should also say that I am reasonably sure I have never encountered these particular warnings/errors. I have accessed this particular VVV site several times over the past few weeks, and have never encountered an error connecting to the database.

[Here](https://pastebin.com/YNb0d4fA) is a pastebin of the complete output from `vagrant provision`.

Maybe as a troubleshooting step, I can step back:  I can undo the `tls-ca` edit I made to `vvv-custom.yml` then try a reprovision to see if the same errors occur?  Do you think that would that be fruitful?

**Comment by tomjn** [comment]
That looks like a new unrelated problem, can you open an issue for it? It looks like a WP CLI problem, have you made any changes to that sites `wp-config.php`?

**Comment by cagross** [comment]
>  have you made any changes to that sites wp-config.php?

Actually, now that you mention it, I did, a few weeks ago.  But I thought I tested everything after the change and confirmed there were no issues.  In hindsight though, I may not have re-provisioned.  Is it necessary to re-provision after making an edit to `wp-config.php`?

Let me look into this some more on my end.  If I can't resolve the issue, I will open a new issue as you suggested.  I'll re-visit the SSL task after.

**Comment by tomjn** [comment]
No, but the error mentioned looked like a PHP issue, and the CA changes didn't touch WP CLI at all. It looks like a database is missing

**Comment by cagross** [comment]
@tomjn Hey I'm back--sorry for the delay.  I was able to resolve that database error myself.  So my `vagrant provision` step completed without issue.  Per your instructions, the next step is:

> If you're using the wordpress-develop or custom site templates, visit your site and switch to `https`.

I am indeed using a custom site template.  But what exactly do you mean by this?  Are you saying I should login to my local WordPress back-end, open Settings-->General, and change both WordPress Address (URL) and Site Address (URL) to `https` ([screenshot](https://imgur.com/6tHYFwI
))?  Or are you referring to another setting--perhaps a VVV setting?

**Comment by tomjn** [comment]
@cagross at this point it's just a standard http -> https transition, nothing VVV specific. It looks like VVV is serving things with https fine, you just need to make WP use it

See this article on the WP specific steps:

https://css-tricks.com/moving-to-https-on-wordpress/

You can skip step 1, that was the VVV certificate authority part

**Comment by tomjn** [comment]
This one uses WP CLI, https://helgeklein.com/blog/2015/01/switching-wordpress-site-http-https/

There's a lot of tutorials/articles out there using different tools to accomplish roughly the same thing

**Comment by cagross** [comment]
OK thanks for that.  I tried following the CSS Tricks article you linked, but was met with an error on the first step (step 2 actually, since we already carried out step 1).  The instruction is to add this line to wp-config:

`define('FORCE_SSL_ADMIN', true);`

I did so, then tried accessing my VVV WordPress back-end (https://cagross.test/wp-admin/).  But I was met with an error:  `NET::ERR_CERT_AUTHORITY_INVALID` ([screenshot](https://imgur.com/7HPrBsY)).  The instructions don't really touch on any potential errors during this step.  They merely say that if all goes well, I will get a secure connection.   Thoughts on that?  It seems to not like the certificate.

**Comment by tomjn** [comment]
Re-read the issue/docs, you have to accept the certificate authority on your OS, else your browser will kick up a fuss the first time it encounters it

**Comment by cagross** [comment]
OK thanks for that.  Yes, my mistake--your previous instructions did indeed mention something like this might happen.  So per your previous instructions, I skipped past that initial 'privacy error' displayed by my browser (Google Chrome).  But when the browser opens my site (back-end or front-end), the navigation bar still displays a 'not secure' error ([screenshot](https://imgur.com/a/gK3Hl)).  Is this expected at this point?  I only ask because the CSS Tricks instructions indicate at this point, the navigation bar should instead display a green 'secure' icon.  Or perhaps your previous instruction regarding certificates/ca/cert.crt will alleviate this issue?  That instruction was:

> Once reprovisioned, you can also go to certificates/ca/cert.crt. Add this to your keychain on MacOS/Accept the certificate as an authority in Windows/etc, and restart your browser. Now all VVV sites using the custom site template support https, with no warnings

I wasn't sure if this was intended to alleviate merely the initial 'privacy error' page, or the navigation bar error as well.

**Comment by tomjn** [comment]
It's kicking up a fuss because it doesn't recognise the custom certificate authority VVV setup to create the certificates. If you accept that authority, it will accept the certificates it creates without question. If you don't accept the certificate, it will warn you every time it encounters a new certificate, and you have to say it's ok each time on a case by case basis

**Comment by cagross** [comment]
OK got it.  So if I understand correctly, my next step is to carry out these instructions:

> you can also go to certificates/ca/cert.crt. Add this to your keychain on MacOS/Accept the certificate as an authority in Windows/etc, and restart your browser.

If so, since I am on Windows, what exactly are the steps here?  I can open the folder certificates/ca, but to be clear, I see a file ca.crt, not cert.crt.  Is ca.crt the file you mean?  And what exactly do you mean by "Accept the certificate as an authority in Windows?"  What does that entail exactly?  Do I need to perhaps add a new SHA to ca.crt?  Or do I need to create a new .crt file?

**Comment by tomjn** [comment]
> If so, since I am on Windows, what exactly are the steps here?

No idea, I'm a Mac user, at this point I'm just acting as a proxy for google

>  I can open the folder certificates/ca, but to be clear, I see a file ca.crt, not cert.crt. Is ca.crt the file you mean?

Probably, I'm mostly going from memory

> And what exactly do you mean by "Accept the certificate as an authority in Windows?" What does that entail exactly?

Certificates come from certificate authorities, which have root certificates. Browsers and operating systems have a list of these root authorities, this way they know if certificates are real or not, because they're signed by those root authorities.

VVV on the other hand simply made one up so it could generate certificates. Your browser sees the SSL certificates, and that it's signed by the VVV authority, and it does not recognise it. Hence the error. You can make it accept that certificate, and it'll pop up again when you add a new site. Or, you can tell your OS about the VVV root certificate ( `ca.cert` ) and it'll know to trust that authority.

How you would do that is platform specific. VVV provides a root certificate, but how you would accept that in your OS is specific to that OS. I don't know the Windows steps, **but you've already done all the VVV specific bits you need to do**. Adding the VVV CA is an optional but recommended step, but you'll have to figure that part out on your own, I can't give a generic catch all platform agnostic answer, and I don't know enough about Windows to be able to give you an answer or know if something I find via Google is correct or not

Likewise, now that your VVV sites have HTTPS certificates, you have to make WP use it. Just as somebody who knows how to speak french and english needs telling to use french if that's what you want them to speak in. You WP site is speaking english, and needs telling to speak french, to use that analogy

**Comment by cagross** [comment]
OK thanks for that explanation.  It helps to illustrate this exact process.

> VVV creates a brand new CA from scratch on provision, that's not been seen before, so you have to tell your OS to trust it.

Am I correct that the certificate issued by VVV's brand new CA is C:\Users\snarl\websites\sites\certificates\ca\ca.cert?

**Comment by tomjn** [comment]
If it isn't it'll kick up a fuss and the adding won't work, it looks correct though

**Comment by cagross** [comment]
OK I got it working.  Here were my steps to add the new VVV certificate to Windows:

* Open the new VVV certificate.  For me, this was named ca.cert, and located in the folder <VVV root>\certificates\ca\ca.cert.
* Follow on-screen prompts to add certificate to Windows.  However, it asks you where you want to install the certificate.  The default option is to allow Windows to choose the location, depending on the type of certificate.  I tried that at first, and it didn't work.  I then tried again, this time specifying a specific location to install:  Trusted Root Certification Authorities store.  I then re-started the browser and the Chrome navigation bar finally displayed the URL as 'secure.'

But now I've observed a new, seemingly related issue.  Every time I restart my browser, when I attempt to login to my VVV WordPress back-end, it forgets my password, i.e. the password I have set for my lone WordPress user isn't recognized.  I have to use the 'forget password' link, then check Mailcatcher for the 'reset password' email and use the link there to reset my password.  This occurs every time I re-start the browser.  If I merely logout of WordPress, then log back in without re-starting the browser, the issue is not present.  I first noticed this issue right after I re-provisioned during the step 'Add tls-ca under core: to the utilities section.'  Since then, I've confirmed it occurs every time I re-start my browser.

Thoughts on this?  Do you want me to perhaps open a new 'issue,' or should we keep going here?

**Comment by tomjn** [comment]
The `tls-ca` utility doesn't touch WP, it only provisions certificates, this is an unrelated issue, and unlikely a VVV one

**Comment by cagross** [comment]
OK got it.  Let me try to do some more troubleshooting on that issue then.  Thanks for your help.

**Comment by cagross** [comment]
@tomjn To update you, to resolve the issue of WordPress forgetting my password, I was advised to run a search-and-replace to replace all `http` URLs with `https` URLs.  I didn't think it would do much, since this was a fresh WordPress install.  But I used a search-and-replace plugin (Better Search Replace) and it returned one result, in the `wp-posts` table.  When I executed the search-replace, the issue went away.  So I guess that was it.  

Thanks for your help in all of that.

**Comment by matwu37** [comment]
@tomjn To follow up on this issue, I'm on VVV 2.1.1, fresh install a site using custom-site-template in `vvv-custom.yml`. Also added `tls-ca` and `trusted-hosts` under the core utilities.

I followed all these steps below and even makes my Win 10 accepted the certificate.

> If you're using the wordpress-develop or custom site templates, visit your site and switch to https. The first time you visit it will complain, but you should be able to skip.

> Once reprovisioned, you can also go to certificates/ca/cert.crt. Add this to your keychain on MacOS/Accept the certificate as an authority in Windows/etc, and restart your browser. Now all VVV sites using the custom site template support https, with no warnings

However, I ran into this weird problem that every custom site seems to be looking at the same `vvv.test` cert, instead of their own DNS name cert. I've attached two images below to showcase, `https://vvv.test` runs fine, whereas `https://local.wordpress.test/` runs into a SSL problem. I tried to edit some vvv-nginx.conf file to force a change without success. Please offer a hand to help me get this SSL problem over with. I've been stuck here for a couple of days. THANK YOU!!

![1](https://user-images.githubusercontent.com/4452309/39969931-e7673eb2-5715-11e8-8701-09adb840a829.PNG)
![2](https://user-images.githubusercontent.com/4452309/39969932-e791187c-5715-11e8-98e2-e41d2472ece8.PNG)

**Comment by tomjn** [comment]
Indeed, if you don't specify the SSL certificate in the nginx config, it'll fallback to the default.

The custom site template does this by putting placeholders in the config template:

https://github.com/Varying-Vagrant-Vagrants/custom-site-template/blob/master/provision/vvv-nginx.conf.tmpl#L12

Then swapping them out in `vvv-init.sh`:

https://github.com/Varying-Vagrant-Vagrants/custom-site-template/blob/master/provision/vvv-init.sh#L57

**Comment by tomjn** [comment]
As an aside:

**If you have an issue with SSL**:

 - Make sure you're running latest versions of everything
 - Double check you're running the latest versions
 - Check the docs
 - Create a new issue
